include/boost/url/impl/any_allocator.hpp

```diff
@@ -0,0 +1,227 @@
+//
+// Copyright (c) 2022 Alan Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_IMPL_ANY_ALLOCATOR_HPP
+#define BOOST_URL_IMPL_ANY_ALLOCATOR_HPP
+
+#include <atomic>
+
+#include <boost/url/detail/over_allocator.hpp>
+#include <boost/type_traits/make_void.hpp>
+#include <boost/type_traits/enable_if.hpp>
+
+namespace boost {
+namespace urls {
+
+// move to new header?
+namespace detail {
+  template<class T, class = void>
+  struct has_is_always_equal : std::false_type {};
+
+  template<class T>
+  struct has_is_always_equal<T, boost::void_t<typename T::is_always_equal>> : std::true_type {};
+
+  // determine if allocator is always equal in a way that works for C++11, C++14, and C++17
+  template <typename A, class Enable = void> struct allocator_is_always_equal : std::is_empty<A> {};
+
+  template <typename A>
+  struct allocator_is_always_equal<A, boost::enable_if_t<has_is_always_equal<A>::value>> : A::is_always_equal {};
+
+  template <typename A>
+  struct allocator_is_always_equal<A, boost::enable_if_t<has_is_always_equal<std::allocator_traits<A>>::value && !std::allocator_traits<A>::value>> : std::allocator_traits<A>::is_always_equal {};
+
+}
+
+
+// base class to store the underlying allocator bound to `base`
+template <class T>
+struct any_allocator<T>::base
+{
+  std::atomic<std::size_t> refs{1};
+
+  virtual
+  ~base()
+  = default;
+
+  explicit
+  base(std::size_t r) : refs(r) {
+
+  }
+
+  virtual
+  char*
+  allocate(std::size_t n) = 0;
+
+  virtual
+  void
+  deallocate(char*p, std::size_t n) = 0;
+
+  template <class U>
+  void
+  deallocate(U* p, std::size_t n) {
+      deallocate(reinterpret_cast<char*>(p), n * sizeof(T));
+  }
+
+  void
+  deallocate(char*p) {
+      deallocate(p, 1);
+  }
+
+  template <class U>
+  void
+  deallocate(U* p) {
+      deallocate(p, 1);
+  }
+
+  virtual void destroy_self() noexcept = 0;
+
+};
+
+template <class T>
+any_allocator<T>::
+~any_allocator()
+{
+  if( p_ && --p_->refs == 0)
+    p_->destroy_self();
+}
+
+template <class T>
+any_allocator<T>&
+any_allocator<T>::
+operator=(
+    any_allocator<T> const& other) noexcept
+{
+  if (this == &other)
+    return *this;
```

Comment:
There's no need for this conditional, it is a pessimization, as almost no assignments will be self-assignments and the code which follows correctly handles this case anyway. Furthermore if this is switched to `shared_ptr` then the entire function disappears. Did you add a test for self-assign?

Reply:
Allowing self-assignment here seems to be a huge problem. The reference counter will be increased twice for the same object and we have a memory leak.

Reply:
look more carefully

Reply:
Oh... OK